File: src/main/java/shortcuts/ConnectedComponents.java
package shortcuts;

public class ConnectedComponents {
    private boolean[] marked;
    private int[] id;
    private int[] size;
    private int count;

    public ConnectedComponents(Graph G) {
        marked = new boolean[G.size()];
        id = new int[G.size()];
        size = new int[G.size()];
        for (int v = 0; v < G.size(); v++) {
            if (!marked[v]) {
                dfs(G, v);
                count++;
            }
        }
    }

    private void dfs(Graph G, int v) {
        marked[v] = true;
        id[v] = count;
        size[count]++;
        G.adj(v).stream().filter(w -> !marked[w]).forEach(w -> dfs(G, w));
    }

    public int id(int v) {
        return id[v];
    }

    public int size(int v) {
        return size[id[v]];
    }

    public int count() {
        return count;
    }

    public boolean connected(int v, int w) {
        return id(v) == id(w);
    }

    public boolean areConnected(int v, int w) {
        return id(v) == id(w);
    }
}

File: src/main/kotlin/shortcuts/ConnectedComponents.kt
package shortcuts

class ConnectedComponents(G: Graph) {
    private val marked = BooleanArray(G.size())
    private val id = IntArray(G.size())
    private val size = IntArray(G.size())
    private var count: Int = 0

    init {
        for (v in 0 until G.size()) {
            if (!marked[v]) {
                dfs(G, v)
                count++
            }
        }
    }

    private fun dfs(G: Graph, v: Int) {
        marked[v] = true
        id[v] = count
        size[count]++
        G.adj(v).stream().filter { w -> !marked[w] }.forEach { w -> dfs(G, w!!) }
    }

    fun id(v: Int): Int {
        return id[v]
    }

    fun size(v: Int): Int {
        return size[id[v]]
    }

    fun count(): Int {
        return count
    }

    fun connected(v: Int, w: Int): Boolean {
        return id(v) == id(w)
    }

    fun areConnected(v: Int, w: Int): Boolean {
        return id(v) == id(w)
    }
}